### ISSUE TITLE
Repeated rebuilds of Hadrian (GHC build system)

### ISSUE THREAD

**Issue by ndmitchell** [commercialhaskell/stack#4566]
Using 

```
$ stack --version
Version 1.10.0, Git revision d1f0fde156245ddbb6656461f3f3531ea84488a2 (dirty) x86_64 hpack-0.31.1
```

Given GHC as checked out with `git clone https://gitlab.haskell.org/ghc/ghc.git --recursive` at SHA 
ef25b59a97f419a2ad6457f696e32aef9ffb3a0f, I do:

```
cd hadrian
stack build hadrian
.. takes a while and works ..
stack build hadrian 2> log.txt
.. rebuilds a large chunk of things and gives lots of error messages ..
```

I would expect that building twice in a row does little to nothing the second time around. Instead, it biulds a whole heap of stuff. The log is attached: [log.txt](https://github.com/commercialhaskell/stack/files/2833450/log.txt)

I note the line `Exception ignored when attempting to load C:\Users\ndmit_000\AppData\Local\Temp\stack7772\hashable-1.2.7.0\.stack-work\dist\7d103d30\stack-config-cache: C:\Users\ndmit_000\AppData\Local\Temp\stack7772\hashable-1.2.7.0\.stack-work\dist\7d103d30\stack-config-cache: openBinaryFile: does not exist (No such file or directory)`

Note sure if that is related.

**Comment by qrilka** [comment]
@ndmitchell we've just today changed the way Stack is supposed to work with GHC boot libraries (in  #4547), and from a quick glance into your log it looks like it could be quite related - Stack complains a lot about mismatching libraries. I'll try to find some time to check out building Hadrian with the latest master - it looks to be a good test case (though a lengthy one )

**Comment by qrilka** [comment]
@ndmitchell it looks like we do some extra rebuilds even using the latest `master` - https://gist.github.com/qrilka/1ee438e352aa8b1539b63bd60483a056 but I get no errors at least.
Checking out if we could have a workaround for boot libraries rebuilds needs further investigation

**Comment by dbaynard** [comment]
Is this resolved, then? It would be good to get an integration test on this. We have `4001-excess-recompilation` from #4003 but (as I understand it) it only checks stack doesn't unregister anything when

1. test/bench dependencies change, and
2. only library is targeted.

**Comment by ndmitchell** [comment]
Definitely not resolved, since I can reproduce it today using Stack from HEAD.

**Comment by qrilka** [comment]
@ndmitchell with fixes in #4588 I see no more repeated rebuilds of Hadrian or its dependencies

**Comment by ndmitchell** [comment]
Confirm it's gone now, so closing.
